src/core/tracker.cpp

```diff
@@ -0,0 +1,162 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QTimer>
+
+#include "tracker.h"
+
+#include "rubberbandmodel.h"
+#include "qgsproject.h"
+
+Tracker::Tracker( QgsVectorLayer *layer, bool visible )
+  : mLayer( layer ),
+    mVisible( visible )
+{
+}
+
+RubberbandModel *Tracker::model() const
+{
+  return mRubberbandModel;
+}
+
+void Tracker::setModel( RubberbandModel *model )
+{
+  if ( mRubberbandModel == model )
+    return;
+  mRubberbandModel = model;
+}
+
+int Tracker::timeInterval() const
+{
+  return mTimeInterval;
+}
+
+void Tracker::setTimeInterval( const int timeInterval )
+{
+  mTimeInterval = timeInterval;
+}
+
+int Tracker::minimumDistance() const
+{
+  return mMinimumDistance;
+}
+
+void Tracker::setConjunction( const bool conjunction )
+{
+  mConjunction = conjunction;
+}
+
+bool Tracker::conjunction() const
+{
+  return mConjunction;
+}
+
+void Tracker::setMinimumDistance( const int minimumDistance )
+{
+  mMinimumDistance = minimumDistance;
+}
+
+void Tracker::trackPosition()
+{
+  if ( std::isnan( model()->currentCoordinate().x() ) || std::isnan( model()->currentCoordinate().y() ) )
+  {
+    return;
+  }
+
+  if ( QgsWkbTypes::hasM( mLayer->wkbType() ) )
+  {
+    if ( !model()->currentPositionTimestamp().isValid() )
+    {
+      return;
+    }
+    double secsSinceStart = 0;
+    if ( mStartPositionTimestamp.isValid() )
+    {
+      secsSinceStart = mStartPositionTimestamp.secsTo( model()->currentPositionTimestamp() );
+    }
+    else
+    {
+      mStartPositionTimestamp = model()->currentPositionTimestamp();
+    }
+    QgsPoint currentCoordinate = model()->currentCoordinate();
+    currentCoordinate.addMValue( secsSinceStart );
+    model()->setCurrentCoordinate( currentCoordinate );
+  }
+
+  model()->addVertex();
+  mTimeIntervalFulfilled = false;
+  mMinimumDistanceFulfilled = false;
+}
+
+void Tracker::positionReceived()
+{
+
+  QVector<QgsPointXY> points = mRubberbandModel->flatPointSequence( QgsProject::instance()->crs() );
+
+  auto pointIt = points.constEnd() - 1;
+
+  QVector<QgsPointXY> flatPoints;
+
+  flatPoints << *pointIt;
+  pointIt--;
+  flatPoints << *pointIt;
+
+  QgsDistanceArea distanceArea;
+  distanceArea.setEllipsoid( QgsProject::instance()->ellipsoid() );
+  distanceArea.setSourceCrs( QgsProject::instance()->crs(), QgsProject::instance()->transformContext() );
+
+  if ( distanceArea.measureLine( flatPoints ) > mMinimumDistance )
```

Comment:
what happens if the project is in lat/lon, does measureLine returns meters?

Reply:
I see we give the tolerance in mapCanvas.mapSettings.destinationCrs.mapUnits. Is it the same?

Reply:
measureLine returns meters, yes. But mapCanvas.mapSettings.destinationCrs.mapUnits returns `deg`. So this need to be fixed...